[MRG] GaussianMixture with BIC/AIC

tingshanL opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/main/CONTRIBUTING.md
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->
Fixes #19338. Automates the selection in [Gaussian Mixture Model Selection](https://scikit-learn.org/stable/auto_examples/mixture/plot_gmm_selection.html). Adds a basic `GaussianMixtureIC` estimator without initializing with agglomerative clustering as discussed in #19562.

#### What does this implement/fix? Explain your changes.
It automatically selects the best GM model based on BIC or AIC among a set of models that are parameterized by:

- Covariance constraints

- Number of components

#### Any other comments?


<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

tliu68 committed:
add basic gmIC

tliu68 committed:
update code

tingshanL committed:
Merge branch 'main' into gmIC

tliu68 committed:
fix linting

tingshanL committed:
Merge branch 'scikit-learn:main' into gmIC

tliu68 committed:
fix linting

tliu68 committed:
fix tests

bdpedigo commented on sklearn/mixture/_gaussian_mixture_ic.py:
```diff
@@ -0,0 +1,360 @@
+"""GaussianMixtureIC"""
+
+# Author: Thomas Athey <tathey1@jhmi.edu>
```
@tingshanL you are the author! truly

bdpedigo commented on sklearn/mixture/_gaussian_mixture_ic.py:
```diff
@@ -0,0 +1,360 @@
+        covariance_type = self._check_parameters()
+        X = self._validate_data(X, dtype=[np.float64, np.float32], ensure_min_samples=1)
+
+        # check n_components against sample size
```
@tingshanL why check both, isnt max > min?

tliu68 committed:
Update _gaussian_mixture_ic.py

tingshanL committed:
Merge branch 'main' into gmIC

tingshanL committed:
Merge branch 'main' into gmIC

tingshanL committed:
Merge branch 'scikit-learn:main' into gmIC

tingshanL committed:
Merge branch 'main' into gmIC

tingshanL committed:
fix docstring

tingshanL committed:
fix attributes

tingshanL committed:
fix attribute typo

tingshanL committed:
fix docstring example mismatch

tingshanL committed:
update docstring example

tingshanL committed:
Merge branch 'main' into gmIC

tingshanL committed:
Merge branch 'main' into gmIC

tingshanL committed:
fix clustering mismatch

tingshanL committed:
Update v1.6.rst

tingshanL committed:
fix linting

tingshanL committed:
fix linting

tingshanL committed:
fix docstring

tingshanL committed:
Update _parameter_constraints

tingshanL committed:
increase codecov

tingshanL committed:
Merge branch 'main' into gmIC

jovo commented:
@amueller @NicolasHug What do you think?

jovo commented:
@jjerphan @ogrisel This PR is designed to greatly simplify user's life when trying to do something like in this tutorial: https://scikit-learn.org/stable/auto_examples/mixture/plot_gmm_selection.html, and is highly simplified relative to other recently proposed PRs, such as https://github.com/scikit-learn/scikit-learn/pull/19562.  We are grateful for your feedback, and look forward to finalizing this.

jovo commented:
@adam2392 Hey Adam - Curious to hear what you think about this?

tingshanL commented:
> @adam2392 Hey Adam - Curious to hear what you think about this?

Hello Adam @adam2392,

We have reviewed and addressed all feedback provided [here](https://github.com/scikit-learn/scikit-learn/pull/19562#pullrequestreview-650296376). Below is a summary of the primary comments and our responses:

- "significantly richer init capabilities than the underlying GaussianMixture class in scikit-learn": currently, `GaussianMixtureIC` selects models only based on covariance constraints and the number of components, the same as the GM model selection [example](https://scikit-learn.org/stable/auto_examples/mixture/plot_gmm_selection.html) in the library.
- "orders of magnitude slower than GaussianMixture": using current `GaussianMixtureIC` to run the same task as in the [example](https://scikit-learn.org/stable/auto_examples/mixture/plot_gmm_selection.html) does not take more time. On my computer, both fitting procedures took around 1.3s.
- "to configure continuous integration to run the tests, including a test to run the [check_estimator](https://scikit-learn.org/stable/modules/generated/sklearn.utils.estimator_checks.check_estimator.html) function to make sure that the code stays compatible with future scikit-learn versions": all checks, including the tests related to `check_estimator`, have passed.
- "joblib-based multi-threading": it is no longer in the code.
- "regularization": also not included in the code anymore.

With these revisions, we believe the code is ready for the next review. Please let us know if there is anything further we should adjust. Thank you very much for your time and insights!

adam2392 commented:
Hey @tingshanL okay thanks! 

I think this will require a discussion among some of the more senior maintainers on the team. Personally, I do see the use of a mclust like algorithm within Python/sklearn ecosystem, since I have used it in the past in R. However, we'll have to see what others say… I know they're busy, and including new models is not super easy, so thank you for the patience.

Out of curiosity (perhaps you can just summarize in the PR description if we want to reserve the space for other discussion), how does this differ from mclust? If there are significant differences, what would need to be done to make this 1-1 matching? This is just some high level info I'm curious on to guide the discussion, so feel free to not spend too much time addressing this question.

tingshanL committed:
Merge branch 'main' into gmIC

tingshanL committed:
Merge branch 'main' into gmIC

tliu68 committed:
fix lint errors

tliu68 committed:
fix linting error

tliu68 committed:
add mahalanobis-ward init

tliu68 committed:
fix lint and pickling errors

tingshanL committed:
Merge branch 'main' into gmIC

adam2392 commented:
Wonder if @scikit-learn/core-devs @scikit-learn/documentation-team have any thoughts? Should we:

A. support this as a "heavy-duty" example in sklearn? - Perhaps this can live in https://scikit-learn.org/stable/auto_examples/index.html. The pro, is we don't maintain the code, and it's an example stringing together 2 and 3 above. Should we also add the different covariance classes too?
B. support this as a class as proposed in this PR, or
C. something else?

Some framings I'm thinking of that could make this more worthwhile maintaining:

- commitment from community to maintain this
- heavy usage of something like this demonstrated already
- overlap of the "mclust-style" covariance types to other parts of the package (i.e. are the covariance structures useful for other models, or simulation classes, etc.)?

adam2392 commented:
Thanks for the work @tingshanL, and apologies if Open Source seems to be slow at times. It is because most (if not all) maintainers are volunteer workers (like me :p). Let's give folks a week or two to chime in.

tingshanL committed:
Merge branch 'main' into gmIC

tliu68 committed:
factor into example

tingshanL committed:
Merge branch 'main' into gmIC

tliu68 committed:
fix linter

tliu68 committed:
Merge branch 'gmIC' of https://github.com/tingshanL/scikit-learn into gmIC

adam2392 commented on doc/whats_new/v1.6.rst:
Can you remove the extra diff in this file?

tliu68 committed:
restore doc
